epoxy-processor/src/main/java/com/airbnb/epoxy/processor/BaseProcessor.kt

```diff
@@ -0,0 +1,166 @@
+        KotlinModelBuilderExtensionWriter(filer, this)
+    }
+
+    override val logger by lazy { Logger(messager, configManager.logTimings) }
+
+    val coroutineExceptionHandler = CoroutineExceptionHandler { _, exception ->
+        logger.logError(exception)
+    }
+
+    override val coroutineScope: CoroutineScope =
+        CoroutineScope(Dispatchers.Default + coroutineExceptionHandler + SupervisorJob())
+
+    override val coroutinesEnabled: Boolean
+        get() = configManager.enableCoroutines
+
+    val generatedModels: MutableList<GeneratedModelInfo> = mutableListOf()
+
+    override fun getSupportedSourceVersion(): SourceVersion = SourceVersion.latestSupported()
+
+    override fun getSupportedAnnotationTypes(): Set<String> =
+        supportedAnnotations().map { it.java.canonicalName }.toSet()
+
+    abstract fun supportedAnnotations(): List<KClass<*>>
+
+    override fun getSupportedOptions(): Set<String> = setOf(
+        ConfigManager.PROCESSOR_OPTION_IMPLICITLY_ADD_AUTO_MODELS,
+        ConfigManager.PROCESSOR_OPTION_VALIDATE_MODEL_USAGE,
+        ConfigManager.PROCESSOR_OPTION_REQUIRE_ABSTRACT_MODELS,
+        ConfigManager.PROCESSOR_OPTION_REQUIRE_HASHCODE,
+        ConfigManager.PROCESSOR_OPTION_DISABLE_KOTLIN_EXTENSION_GENERATION,
+        ConfigManager.PROCESSOR_OPTION_LOG_TIMINGS,
+        ConfigManager.PROCESSOR_OPTION_ENABLE_PARALLEL
+    )
+
+    override fun init(processingEnv: ProcessingEnvironment) {
+        super.init(processingEnv)
+
+        filer = processingEnv.filer
+        messager = processingEnv.messager
+        elementUtils = processingEnv.elementUtils
+        typeUtils = processingEnv.typeUtils
+        configManager = ConfigManager(processingEnv.options, elementUtils, typeUtils)
+        synchronizationEnabled = configManager.enableCoroutines
+        resourceProcessor = ResourceProcessor(processingEnv, logger, elementUtils, typeUtils)
+    }
+
+    private var roundNumber = 1
+
+    final override fun process(
+        annotations: Set<TypeElement?>,
+        roundEnv: RoundEnvironment
+    ): Boolean = runBlocking(Dispatchers.Default) {
+        try {
+            logger.measure("Process Round: $roundNumber") {
+                processRound(roundEnv, roundNumber++)
+            }
+        } catch (e: Exception) {
+            logger.logError(e)
+        }
+
+        if (roundEnv.processingOver()) {
+            val processorName = this@BaseProcessor::class.java.simpleName
+            // We wait until the very end to log errors so that all the generated classes are still
+            // created.
+            // Otherwise the compiler error output is clogged with lots of errors from the generated
+            // classes  not existing, which makes it hard to see the actual errors.
+            logger.measure("validateAttributesImplementHashCode") {
+                validateAttributesImplementHashCode(generatedModels)
+            }
+
+            logger.writeExceptions()
```

Comment:
Probably not necessary but just a thought - it seems like normal exceptions are pretty tractable and likely can be gracefully lazily collected. 

Did you find any super weird output with concurrent exceptions like cme's while working on the parallelization? I could imagine that those might leave the program in a weird state and create tons of exceptions. Specially handling them might also be nice to print a message like "CME found, consider disabling experimental parallelization feature and file a ticket with info).

Reply:
interesting idea. I did find `CME` sometimes, or a specific exception from ClassReader. i like the idea of checking for that to tell users to disable parallelization and to report the stacktrace. i didn't find any times where the processor got into a weird state because of it where there were downstream errors